mk/golang.mk

```diff
@@ -12,7 +12,8 @@ GOTFLAGS ?=
 unexport GOFLAGS
 # Override so we can combine with the user's go flags.
 # Try to make building as reproducible as possible by stripping the go path.
-override GOFLAGS += "-asmflags=all='-trimpath=$(GOPATH)'" "-gcflags=all='-trimpath=$(GOPATH)'"
```

Comment:
There be balrogs here.

I need to be able to:

1. Checkout the go-ipfs source locally.
2. Run `make install`.
3. Checkout the https://github.com/ipfs/go-ds-s3/ reop.
4. Run `make install IPFS_VERSION=/path/to/go-ipfs/source` (s3 datastore plugin).
5. Start the IPFS daemon, correctly loading the s3 plugin.

Unfortunately, we've found that:

1. The bare `-trimpath` flag works with plugins when go-ipfs is built with `go build github.com/...`.
2. The bare `-trimpath` flag _does not_ work with plugins when go-ipfs is built with `make build`.

Somewhere in the deep dark reaches of Khazad-dûm (the go compiler), the plugin loader gets confused.

If you can figure out how to reconcile this... I'll be very happy.

Reply:
Looking into this now.

Some interesting things I found when building without bare `-trimpath`.

The only difference between binaries is this path not being trimmed.

```
/tmp/go-build971257389/b082/_cgo_gotypes.go
```

I'm guessing this is another issue with the compiler.  Maybe an easier one to fix than the plugin loader.

Reply:
Note: I haven't checked this for at least one go release cycle. The situation may have improved.

Reply:
It seems the situation has not improved https://github.com/golang/go/issues/31354.

Building the plugins within the same path as the application seems to be a viable option.

[This](https://github.com/lf-edge/ekuiper/blob/master/docs/en_US/plugins/overview.md) project is able to build plugins with `-trimpath` via this method.

Reply:
I've managed to get the go-ds-s3 plugin working by building inside of the IPFS project root.

Steps I've taken to produce the binary and plugin:

Clone `go-ds-s3` repo to `extensions` folder.

Copy `go.mod` to `extensions.mod`

Copy `go.sum` to `extensions.sum`

Edit `extensions.mod` with the following:

```makefile
require (
	...
	github.com/ipfs/go-ds-s3 v0.0.0
)

replace github.com/ipfs/go-ds-s3 => ./extensions/go-ds-s3
```

Edit `extensions/go-ds-s3/go.mod` with the following:

```makefile
replace github.com/ipfs/go-ipfs => ../../
```

Copy line below from `extensions/go-ds-s3/go.sum` to `extensions.sum`

```makefile
github.com/ipfs/go-ipfs v0.8.0/go.mod h1:iU9jeKJrinNYdWSNZyTjpfcmGHH4oflEySOUboEQsoQ=
```

Build go-ipfs with `-trimpath`

```makefile
GOFLAGS=-trimpath make build
```

Build go-ds-s3 plugin with `-trimpath`

```makefile
go build -trimpath -modfile extensions.mod --buildmode=plugin -o s3plugin.so extensions/go-ds-s3/plugin/main/main.go
```

Reply:
That's a pretty big step backwards in terms of usability. Plugins are already confusing.

Reply:
What happens if you use `-trimpaths` _and_ the asm/gcflags?

Reply:
Another acceptable solution would be to make the Makefile _not_ override user-provided flags, but instead set defaults.

Reply:
> That's a pretty big step backwards in terms of usability. Plugins are already confusing.

I totally agree.  My goal was to first find a solution and then work on developer ergonomics.  Possibly a Dockerfile with everything setup or a script that automates the steps above.  I'm hoping to move everything to https://github.com/ipfs/distributions once I find an acceptable solution.

> What happens if you use `-trimpaths` _and_ the asm/gcflags?

Using the asm/gcflags somehow overrides `-trimpath` and causes the build to be non reproducible.  The Go team suggests to avoid using those flags and use `-trimpath` instead.

> Another acceptable solution would be to make the Makefile _not_ override user-provided flags, but instead set defaults.

I think this is a good solution.  I wasn't able to successfully set `-ldflags` via `GOFLAGS` but that is probably my lack of experience with Makefiles.

Reply:
I've managed to reduce the steps as well as allow building from an outside directory.

Clone `go-ipfs` and `go-ds-s3` to the same directory.

```bash
git clone -b v0.8.0 https://github.com/ipfs/go-ipfs.git
git clone -b v0.7.0 https://github.com/ipfs/go-ds-s3.git
``` 

Edit the `./go-ds-s3/go.mod`.

```bash
go mod edit -replace "github.com/ipfs/go-ipfs=../go-ipfs"
```

Edit the `./go-ipfs/go.mod`.

```bash
go mod edit -require="github.com/ipfs/go-ds-s3@v0.7.0"
go mod edit -replace "github.com/ipfs/go-ipfs=../go-ds-s3"
go mod tidy
# I'm unsure why this is required but possibly due to nested directory structure
go get github.com/ipfs/go-ds-s3/plugin
```

Comment out this line in `./go-ipfs/mk/golang.mk`.

```Makefile
#override GOFLAGS += "-asmflags=all='-trimpath=$(GOPATH)'" "-gcflags=all='-trimpath=$(GOPATH)'"
```

Build `ipfs` from `./go-ipfs`:

```bash
GOFLAGS=-trimpath make build
```

Build `go-ds-s3` from `./go-ipfs`:

```bash
go build -trimpath --buildmode=plugin -o go-ds-s3.so ../go-ds-s3/plugin/main/main.go
```